src/id-generation.ts

```diff
@@ -0,0 +1,61 @@
+import {nanoid} from "npm:nanoid@4.0.0"
+import {
+    createHash
+} from "https://deno.land/std@0.168.0/node/internal/crypto/hash.ts"
+
+
+export type IdGenerator = (data?: string) => string
+
+
+export const nanoIdGen = (): IdGenerator =>
+    () => nanoid()
+
+
+export const sequentialIdGen = (): IdGenerator => {
+    let num = 0
+    return () => `${++num}`
+}
+
+
+const defaultHashAlgorithm = "SHA256"
+
+export type IdGenOptions = {
+    doNotCheckForUniqueData?: boolean
+    algorithm?: typeof defaultHashAlgorithm | string
+    salt?: string
+    checkForUniqueHash?: boolean
+}
+
+export const hashingIdGen = (options?: IdGenOptions): IdGenerator => {
+    const checkForUniqueData = !(options?.doNotCheckForUniqueData)
+    const algorithm = options?.algorithm ?? defaultHashAlgorithm
+    const salt = options?.salt ?? ""
+    const checkForUniqueHash = options?.checkForUniqueHash
+
+    const datas: string[] = []
+    const hashes: string[] = []
+
+    return (data?: string) => {
+        if (data === undefined) {
+            throw new Error(`expected data for hashing`)
+        }
+        if (checkForUniqueData) {
+            if (datas.indexOf(data) > -1) {
+                throw new Error(`duplicate data encountered: "${data}"`)
+            }
+            datas.push(data)
+        }
+        const hash = createHash(algorithm)
+            .update(salt + data)
```

Comment:
What is the advantage of using salt here? I am very ignorant of this topic, so I may miss something obvious. I thought that salt was useful to make more difficult to decrypt something, but I think this is not a goal for us in this context

Reply:
The advantage is to be able to make hashed IDs unique across models even if you e.g. hash a qualified name. Qualified names (within metamodels) might be the same across models, and would produce the same ID without a salt. Of course, the salt would have to be different... Maybe we can remove this again when we've solved issue #19 .